pkg/workflow/controller/controllers/controller.go

```diff
@@ -107,17 +107,30 @@ func (c *Controller) doWork(key string) error {
 		return fmt.Errorf("Error fetching object with key %s from store: %v", key, err)
 	}
 
-	if !exists {
-		deleteObj, err := c.drCollection.Get(key)
-		if err != nil {
-			log.WithField("key", key).Error("Deleted object lost")
-			return nil
-		}
-		if err := c.eventHandler.ObjectDeleted(deleteObj); err != nil {
-			return err
-		}
-		c.drCollection.Remove(key)
+	if obj == nil || !exists {
+		log.WithField("obj", obj).WithField("exist", exists).Warning("Object is nil or not exist")
 		return nil
 	}
+
+	object, ok := obj.(metav1.Object)
+	if !ok {
+		log.WithField("obj", obj).Warning("Expect it is a Kubernetes resource object, got unknown type resource")
+		return fmt.Errorf("unknown resource type")
+	}
+
+	if c.isBeingDeleted(object) {
+		return c.eventHandler.HandleFinalizer(object)
+	}
+
+	// Add finalizer if needed
+	if err := c.eventHandler.AddFinalizer(object); err != nil {
+		return err
+	}
+
 	return c.eventHandler.Reconcile(obj)
 }
+
+// isBeingDeleted indicates whether the resource is being deleted
+func (c *Controller) isBeingDeleted(obj metav1.Object) bool {
```

Comment:
This is not a method of `Controller`. As its logic is very simple, I'd prefer to not define this method if it is not widely called.

Reply:
ok